### PR TITLE
Examples Browser

### PR DESCRIPTION
**Pull request by ellthompson**
This PR rewrites the examples browser as a TypeScript and React based single page application. Examples can be edited and rerun on desktop browsers using the Monaco code editor.

The readme has been updated with instructions for how to locally build the app and contribute new examples.

image

image

Local usage:
- Go to `./examples`
- `npm install`

Then in three separate terminals

- `npm run build:watch`
- `npm run serve`
- `npm run thumbnails`

Todo before release:
- [x] Play button
- [x] Embed / Share functionality
- [x] Refactor
- [x] Copy over all examples
- [x] Readme
- [x] iframe directory
- [x] non hashed example paths

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by willeastcott** [comment]
Did you forget to check in `thumbnails.js`?

**Comment by ellthompson** [comment]
> Did you forget to check in `thumbnails.js`?

It's there but it's location has changed. I'll update the package.json

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:32]:
```suggestion
To create a new example you can copy any of the existing examples as a template and update its path.
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:34]:
```suggestion
Each example extends the `Example` parent class and can implement three methods to define its functionality:
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:43]:
```suggestion
This is the only function that's required to run an example. The code defined in this function is executed each time the example play button is pressed. It takes the example's canvas element as its first argument and usually begins by creating a new PlayCanvas application using that canvas.
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:57]:
```suggestion
As assets are loaded using React, be sure to import React into any example that is loading assets.
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:81]:
```suggestion
Each script will be made available as a parameter of the example function as an esModule using the name it was given and therefore any scripts should be defined in the examples function signature as follows:
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:60]:
```suggestion
Each asset you load will be made available to the `example` function you write as the second parameter and will already be in the loaded state.
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:90]:
```suggestion
This function allows you to define a set of PCUI based interface which can be used to display stats from your example or provide users with a way of controlling the example.
```

File: examples/README.md
Comment (willeastcott) [inline examples/README.md:101]:
```suggestion
The controls function takes a [pcui observer](https://playcanvas.github.io/pcui/data-binding/using-observers/) as its parameter and returns a set of PCUI components. Check this [link](https://playcanvas.github.io/pcui/examples/todo/) for an example of how to create and use PCUI.
```

File: examples/package.json
Comment (willeastcott) [inline examples/package.json:81]:
```suggestion
  "author": "PlayCanvas <support@playcanvas.com>",
  "license": "MIT"
```

File: examples/package.json
Comment (willeastcott) [inline examples/package.json:4]:
```suggestion
"description": "Examples browser for the PlayCanvas Engine",
```

File: examples/src/app/example-iframe.tsx
Comment (willeastcott) [inline examples/src/app/example-iframe.tsx:76]:
```suggestion
        // strip the PlayCanvas app initialization
```

File: examples/src/app/example-iframe.tsx
Comment (willeastcott) [inline examples/src/app/example-iframe.tsx:100]:
```suggestion
        // Set the canvas to fill the window and automatically change resolution to be the same as the canvas size
```

**Review by willeastcott — APPROVED** [review]

**Comment by mvaligursky** [comment]
Look great!
Some comments based on just using the example browser:

- the mini stats is missing (few examples have relevant sections commented out, for example clustered-lighting).
Screenshot 2021-06-09 at 11 51 58

- some examples which were not exposed before (private, waiting for release) are now public (example: Lights Baked AO). Is there some way to hide it?
- in my list of examples, I have example "Lights" twice .. I believe the second one used to have different name
Screenshot 2021-06-09 at 11 13 28

- with small thumbnails enabled, I find it pretty hard to quickly see the section name (Camera / Graphics ...) when I just quickly want to find them .. we could make those more stand out. Similarly, we could more highlight the currently running example too. This was clearer in the previous example browser.
Screenshot 2021-06-09 at 11 54 23

- I'm not sure about Editor vs preview proportions .. sometimes the preview is tiny compared to the Editor.
Screenshot 2021-06-09 at 11 12 56
- mini-stats and multi-application example are missing - can we have those back
- open an example with longer code, scroll to the end of the code, and then switch to hello-world. The code is scrolled out and not visible. Perhaps reset editor to display the start of the example when switching examples.
- open point cloud example, edit vertex shader (for example change the color R value on last line). Press play - the change is reflected. Then switch to another example, and see the error:
Screenshot 2021-06-09 at 11 28 12
- when you click on the same sample multiple times, you get this in the console:
Screenshot 2021-06-09 at 11 29 57
- post-effects has warnings:
Screenshot 2021-06-09 at 11 34 05
- looking specifically at render-cubemap. The render-cubemap.tsx has the section on top which loads assets [helipad.dds, cubemap-renderer.js] but that section of the code is not visible in the code preview. The code preview simply creates a component using it: 
` shinyBall.script.create('cubemapRenderer', {`
I don't like that this hides the need for the user to load the script. They'll copy and paste the code to their project and it won't work. Perhaps we need at least a link to full example source code in github? Or maybe this script should be added as a tab in the code editor? Or even better - a list of resources to load should be visible in the Editor view.

**Comment by mvaligursky** [comment]
also, this folder should probably be excluded from git so it does not show up as changes
Screenshot 2021-06-09 at 13 53 46

**Comment by ellthompson** [comment]
> also, this folder should probably be excluded from git so it does not show up as changes
> Screenshot 2021-06-09 at 13 53 46

This is listed in the .gitignore file but unfortunately as this isn't merged into master yet it isn't picked up correctly. The issue should resolve once this PR is merged.

File: examples/src/app/helpers/raw-file-loading.ts
Comment (mvaligursky) [inline examples/src/app/helpers/raw-file-loading.ts:1]: some of the files have no comments at all .. it'd be good to add at least some description of the purpose of the file / class.

**Comment by mvaligursky** [comment]
One more question - can this be used with the automated testing @slimbuck  set up?

**Comment by mvaligursky** [comment]
I got this, but not sure how to reproduce. I was looking at the Point Clouds, and tabs in the editor part started flickering and then the error.
Screenshot 2021-06-09 at 18 02 20

**Comment by mvaligursky** [comment]
if easy to do, we could hide these type of lines from showing up in the editor
`// @ts-ignore engine-tsd`

File: examples/src/examples/graphics/render-cubemap.tsx
Comment (mvaligursky) [inline examples/src/examples/graphics/render-cubemap.tsx:149]: just curious about why this line got so complicated?

File: examples/src/examples/graphics/render-cubemap.tsx
Comment (mvaligursky) [inline examples/src/examples/graphics/render-cubemap.tsx:227]: have you integrated master to this branch in last few weeks? I think the syntax on left with options parameters is the last one, not passing params as here. Worth checking.

**Comment by mvaligursky** [comment]
I looked at the code changes now, and apart from two small comments it all looks great - awesome work!

File: examples/src/examples/graphics/render-cubemap.tsx
Comment (ellthompson) [reply examples/src/examples/graphics/render-cubemap.tsx:149]: Oh this was created by VSCode's autotyping for TypeScript. I guess it inferred that entities each contain a rotate method (used on line 215). I'll remove this and just keep the pc.Entity[] type.

File: examples/src/examples/graphics/render-cubemap.tsx
Comment (ellthompson) [reply examples/src/examples/graphics/render-cubemap.tsx:227]: Oh yeah, i'll bump the version of pc to 1.42.0

Comment (mvaligursky) [reply examples/src/examples/graphics/render-cubemap.tsx:227]: ah that's how it works. Will it work fine to bump it to a version we're about to release to release examples at the same time? Or will we release new version of the engine, then bump up version here and then release engine examples?

**Comment by Maksims** [comment]
By implementing all examples in TypeScript, encourages new users to use TypeScript.
**This definitely requires a bigger consideration.**

File: examples/src/examples/misc/mini-stats.tsx
Comment (mvaligursky) [inline examples/src/examples/misc/mini-stats.tsx]:
```suggestion
        // Note: for most of these to report values, either debug or profiling engine build needs to be used.
```

**Comment by mvaligursky** [comment]
> By implementing all examples in TypeScript, encourages new users to use TypeScript.

The plan here is to have a switch in the examples app between TS and JS. We write them in TS, but users can use either, and TS would get transpiled to JS. Is this still the case @ellthompson ?

**Comment by Maksims** [comment]
> > By implementing all examples in TypeScript, encourages new users to use TypeScript.
> 
> The plan here is to have a switch in the examples app between TS and JS. We write them in TS, but users can use either, and TS would get transpiled to JS. Is this still the case @ellthompson ?

I just looking at a screenshot in first post.
Also, is sandboxing (remixing) of examples code will be supported, or source will be read-only?

**Comment by mvaligursky** [comment]
> Also, is sandboxing (remixing) of examples code will be supported, or source will be read-only?

Yes, that works already, I was testing that yesterday .. you can change code, and hit play. I'm not sure what happens if I share this to twitter .. is this going to share my changes or original example? @ellthompson 

Perhaps it'd be good to improve the description of the PR of existing functionality, and link to an issue with future enhancements that are planned.

**Comment by ellthompson** [comment]
Yeah the plan is to have a toggle to switch between TS and JS. However i'm still researching options on how to implement this so it won't make the first release.

Sandboxing is currently supported. You can edit examples in the code pane and hit the play button to execute them again.

**Comment by ellthompson** [comment]
I'm going ahead and merging this now. I'll create github issues for outstanding tasks (Displaying loaded scripts as separate files / TS to JS toggle).

**Pull request merged**
